cep-16.md

```diff
@@ -0,0 +1,260 @@
+
+However, the file still contains all repodata in the channel. This means the file needs to be redownloaded every time anyone adds a single package (even if a user doesnt need that package).
+
+Because the file is relatively big this means that often a large `max-age` is used for caching which means it takes more time to propagate new packages through the ecosystem.
+
+## Proposal
+
+We propose a "sharded" repodata format. It works by splitting the repodata into multiple files (one per package name) and recursively fetching the "shards".
+
+The shards are stored by the hash of their content (e.g. "content-addressable"). 
+That means that the URL of the shard is derived from the content of the shard. This allows for efficient caching and deduplication of shards on the client. Because the files are content-addressable no round-trip to the server is required to check freshness of individual shards.
+
+Additionally an index file stores the mapping from package name to shard hash.
+
+Although not explicitly required the server SHOULD support HTTP/2 to reduce the overhead of doing a massive number of requests. 
+
+### Repodata shard index
+
+The shard index is a file that is stored under `<shard_base_url>/<subdir>/repodata_shards.msgpack.zst`. It is a zstandard compressed `msgpack` file that contains a mapping from package name to shard hash. The `<shard_base_url>` is defined in [Authentication](#authentication).
+
+The contents look like the following (written in JSON for readability):
+
+```js
+{
+  "version": 1,
+  "info": {
+    "base_url": "https://example.com/channel/subdir/",
+    "created_at": "2022-01-01T00:00:00Z",
+    "...": "other metadata"
+  },
+  "shards": {
+    // note that the hashes are stored as binary data (hex encoding just for visualization)
+    "python": b"ad2c69dfa11125300530f5b390aa0f7536d1f566a6edc8823fd72f9aa33c4910",
+    "numpy": b"27ea8f80237eefcb6c587fb3764529620aefb37b9a9d3143dce5d6ba4667583d"
+    "...": "other packages"
+  }
+}
+```
+
+The index is still updated regularly but the file does not increase in size with every package added, only when new package names are added which happens much less often.
+
+For a large case (conda-forge linux-64) this files is 670kb at the time of writing.
+
+We suggest serving the file with a short lived `Cache-Control` `max-age` header of 60 seconds to an hour but we leave it up to the channel administrator to set a value that works for that channel.
+
+### Repodata shard
+
+Individual shards are stored under the URL `<shard_base_url>/<subdir>/shards/<sha256>.msgpack.zst`. Where the `sha256` is the lower-case hex representation of the bytes from the index. It is a zstandard compressed msgpack file that contains the metadata of the package. The `<shard_base_url>` is defined in [Authentication](#authentication).
```

Comment:
What if we got rid of the implicit /shards/ subdir?

Reply:
Any reasons for getting rid of it?

Reply:
There are some benefits I can think of:

- makes it easy for the backend to redirect traffic to another bucket or other rewrite rules
- separates the data, makes it easy to purge the whole `shards/` folder if necessary
- doesn't clutter the folder with the packages

But I would be happy to hear your reasons for removing it.

Reply:
It's something else to look up or remember and we have some upload code that assumes everything for a subdir is in the same folder. Should the shards folder be part of the index file?

According to google some people call .msgpack .mpk but others don't want to advertise "generic msgpack" as a thing and instead use an application extension.

Reply:
@jezdez was asking to not optimize for a specific way of distributing these things :) some people might still run on traditional filesystems ...

Reply:
I could mention splitting the hash into pairs and having a tree of ab/cd/ef/abc123.zst
But we probably have to assume that our filesystems are OK with large numbers of files in the same folder since the packages are there.

Reply:
Is there a reason not to include "shard_base_url": "https://shards.prefix.dev/conda-forge/<subdir>/shards/" (including a now-explicit /shards/ subdirectory) in the index file?

Reply:
Nope, I think that makes more sense indeed.

Reply:
I added this to the CEP, implemented it [in our mirror script](https://github.com/prefix-dev/sharded-repo/pull/6), and added [this to rattler](https://github.com/mamba-org/rattler/pull/747).